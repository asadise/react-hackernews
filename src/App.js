import React, { Component } from 'react';
import Search from './Search.js'
import Table from './Table.js'
//import logo from './logo.svg';
import './App.css';

const list = [
  {
    title: 'صد سال تنهایی',
    url: 'https://facebook.github.io/react/',
    author: 'گابریل مارکز',
    num_comments: 3,
    points: 4,
    objectID: 0,
  },
  {
    title: 'جنگ و صلح',
    url: 'https://github.com/reactjs/redux',
    author: 'تلستوی',
    num_comments: 2,
    points: 5,
    objectID: 1,
  },
];

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      list: list,
      searchTerm: '',
    }
    this.onDismiss = this.onDismiss.bind(this);
    this.onChangeSearch = this.onChangeSearch.bind(this);
  }

  onChangeSearch(event) {
    this.setState({ searchTerm: event.target.value });
  }

  onDismiss(id) {
    const isNot = item => item.objectID !== id;
    const upldatedList = this.state.list.filter(isNot);
    /* const upldatedList = this.state.list.filter(function (listItem) {
      return listItem.objectID !== id;
    }); */
    this.setState({ list: upldatedList })
  }

  render() {
    let hi = 'سلام به اپ ری‌اکت من خوش آمدید!';
    const { list, searchTerm } = this.state;

    return (
      <div className="App" dir="rtl">
        <h2>{hi}</h2>
        <Search value={searchTerm} onChange={this.onChangeSearch} />
        <Table list={list} pattern={searchTerm} onDismiss={this.onDismiss} />
      </div>
    );
  }
}

export default App;
