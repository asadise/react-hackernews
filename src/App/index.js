import React, { Component } from 'react';
import axios from 'axios'
import Search from '../Search'
import Table from '../Table'
import LoadingImage from '../resource/loading.gif'
import './index.css';

const PATH_BASE = 'http://my-json-server.typicode.com/asadise/book-api';
const PARAM_SEARCH = '/books';
const PARAM_PAGE = '_limit=2&_page=';

function isEmpty(obj) {
  return !obj || Object.keys(obj).length === 0;
}

class App extends Component {
  _isMounted = false;
  constructor(props) {
    super(props);
    this.state = {
      result: null,
      searchTerm: '',
      page: 1,
      error: null,
      isLoading: false,
    }
    this.onDismiss = this.onDismiss.bind(this);
    this.onSearchSubmit = this.onSearchSubmit.bind(this);
    this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this);
    this.setSearchTopStories = this.setSearchTopStories.bind(this);
    this.onChangeSearch = this.onChangeSearch.bind(this);
  }

  onSearchSubmit(event) {
    const { searchTerm } = this.state;
    this.fetchSearchTopStories(searchTerm);
    event.preventDefault();
  }

  fetchSearchTopStories(searchTerm = '', page = 1) {
    this.setState({ isLoading: true });
    const url = (searchTerm) ? `${PATH_BASE}${PARAM_SEARCH}/${searchTerm}` : `${PATH_BASE}${PARAM_SEARCH}?${PARAM_PAGE}` + page;
    axios(url)
      .then(result => this._isMounted && this.setSearchTopStories(result.data, searchTerm, page))
      .catch(error => this._isMounted && this.setState({ error }));
  }

  setSearchTopStories(result, searchTerm, page) {
    let oldResult = (this.state.result) ? this.state.result : [];
    if (searchTerm !== '') {
      oldResult = [];
      page = -1;
    }
    //if result is an object instead of array
    let resultArray = [];
    if (!isEmpty(result) && result.length === undefined)
      resultArray.push(result);
    else if (!isEmpty(result))
      resultArray = result;

    const upldatedResult = [...oldResult, ...resultArray];
    this.setState({ result: upldatedResult, searchTerm: searchTerm, page: page, isLoading: false })
  }

  onChangeSearch(event) {
    this.setState({ searchTerm: event.target.value });
  }

  componentDidMount() {
    this._isMounted = true;
    this.fetchSearchTopStories();
  }

  onDismiss(id) {
    const isNot = item => item.id !== id;
    const upldatedResult = this.state.result.filter(isNot);
    /* const upldatedList = this.state.list.filter(function (listItem) {
      return listItem.id !== id;
    }); */
    /*this.setState({ result: Object.assign({}, this.state.result, { books: upldatedResult }) })
      const userNames = { firstname: 'Robin', lastname: 'Wieruch' }; const userAge = { age: 28 };
      const user = { ...userNames, ...userAge };
    */
    this.setState({ result: upldatedResult })
  }

  render() {
    let hi = 'سلام؛ به پروژه «هکر نیوز» در React خوش آمدید!';
    const { result, searchTerm, page, error, isLoading } = this.state;

    if (error)
      return <div className="App" style={{ textAlign: "center" }}><h3>خطایی در فراخوانی وب‌سرویس رخ داده است.</h3><span>جزئیات: {error.toString()}</span></div>

    return (
      <div className="App">
        <div className="page">
          <div className="intersection">
            <h2>{hi}</h2>
            <Search value={searchTerm} onChange={this.onChangeSearch} onSubmit={this.onSearchSubmit} >جستجو...</Search>
            {result ? <Table list={result} onDismiss={this.onDismiss} fetchSearchTopStories={this.fetchSearchTopStories} page={page} isLoading={isLoading} /> : null}
            {isLoading ? <Loading /> : ""}
          </div>
        </div>
      </div>
    );
  }
}

const Loading = () =>
  <div style={{ textAlign: "center" }}>
    <img src={LoadingImage} alt="در حال بارگزاری؛ لطفا صبر کنید..." />
  </div>

export default App;
