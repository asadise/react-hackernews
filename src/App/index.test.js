import React from 'react';
import ReactDOM from 'react-dom';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import App from '.';
Enzyme.configure({ adapter: new Adapter() });

describe('App', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });
  test('has a valid snapshot', () => {
    const component = renderer.create(<App />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

test('renders learn react link', () => {
  const { getByText } = render(<App />);
  const linkElement = getByText(/React/i);
  expect(linkElement).toBeInTheDocument();
});


