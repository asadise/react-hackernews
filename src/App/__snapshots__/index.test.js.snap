// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App has a valid snapshot 1`] = `
<div
  className="App"
>
  <div
    className="page"
  >
    <div
      className="intersection"
    >
      <h2>
        سلام؛ به پروژه «هکر نیوز» در React خوش آمدید!
      </h2>
      <div>
        <form
          onSubmit={[Function]}
        >
          <input
            onChange={[Function]}
            placeholder="جستجو..."
            type="text"
            value=""
          />
           
          <button
            type="submit"
          >
            جستجو...
          </button>
        </form>
        <br />
        <br />
        <br />
      </div>
    </div>
  </div>
</div>
`;
